storage/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogManagerConfig.java

```diff
@@ -348,280 +346,124 @@ public final class RemoteLogManagerConfig {
-                && Objects.equals(remoteStorageManagerClassName, that.remoteStorageManagerClassName)
-                && Objects.equals(remoteStorageManagerClassPath, that.remoteStorageManagerClassPath)
-                && Objects.equals(remoteLogMetadataManagerClassName, that.remoteLogMetadataManagerClassName)
-                && Objects.equals(remoteLogMetadataManagerClassPath, that.remoteLogMetadataManagerClassPath)
-                && Objects.equals(remoteLogMetadataManagerListenerName, that.remoteLogMetadataManagerListenerName)
-                && Objects.equals(remoteStorageManagerProps, that.remoteStorageManagerProps)
-                && Objects.equals(remoteLogMetadataManagerProps, that.remoteLogMetadataManagerProps)
-                && Objects.equals(remoteStorageManagerPrefix, that.remoteStorageManagerPrefix)
-                && Objects.equals(remoteLogMetadataManagerPrefix, that.remoteLogMetadataManagerPrefix)
-                && remoteLogManagerCopyMaxBytesPerSecond == that.remoteLogManagerCopyMaxBytesPerSecond
-                && remoteLogManagerCopyNumQuotaSamples == that.remoteLogManagerCopyNumQuotaSamples
-                && remoteLogManagerCopyQuotaWindowSizeSeconds == that.remoteLogManagerCopyQuotaWindowSizeSeconds
-                && remoteLogManagerFetchMaxBytesPerSecond == that.remoteLogManagerFetchMaxBytesPerSecond
-                && remoteLogManagerFetchNumQuotaSamples == that.remoteLogManagerFetchNumQuotaSamples
-                && remoteLogManagerFetchQuotaWindowSizeSeconds == that.remoteLogManagerFetchQuotaWindowSizeSeconds;
+    public RemoteLogManagerConfig(Map<?, ?> props) {
+        this(props, false);
     }
 
-    @Override
-    public int hashCode() {
-        return Objects.hash(enableRemoteStorageSystem, remoteStorageManagerClassName, remoteStorageManagerClassPath,
-                            remoteLogMetadataManagerClassName, remoteLogMetadataManagerClassPath, remoteLogMetadataManagerListenerName,
-                            remoteLogMetadataCustomMetadataMaxBytes, remoteLogIndexFileCacheTotalSizeBytes, remoteLogManagerThreadPoolSize,
-                            remoteLogManagerCopierThreadPoolSize, remoteLogManagerExpirationThreadPoolSize, remoteLogManagerTaskIntervalMs,
-                            remoteLogManagerTaskRetryBackoffMs, remoteLogManagerTaskRetryBackoffMaxMs, remoteLogManagerTaskRetryJitter,
-                            remoteLogReaderThreads, remoteLogReaderMaxPendingTasks, remoteStorageManagerProps, remoteLogMetadataManagerProps,
-                            remoteStorageManagerPrefix, remoteLogMetadataManagerPrefix, remoteLogManagerCopyMaxBytesPerSecond,
-                            remoteLogManagerCopyNumQuotaSamples, remoteLogManagerCopyQuotaWindowSizeSeconds, remoteLogManagerFetchMaxBytesPerSecond,
-                            remoteLogManagerFetchNumQuotaSamples, remoteLogManagerFetchQuotaWindowSizeSeconds);
+    protected RemoteLogManagerConfig(Map<?, ?> props, boolean doLog) {
```

Comment:
This constructor isn't doing anything. You can call super(ConfigDef, Map) in the other constructor.

Also don't forget to remove CONFIG_DEF and replace it with a ConfigDef configDef() static method.

Reply:
It seems to me composition is better than inheritance. Extending `AbstractConfig` may add many unrelated/unexpected methods to `RemoteLogManagerConfig`. Being a POJO is more safe. 

Hence, maybe we can keep a inner `AbstractConfig` object? All getters can be re-implemented by that inner `AbstractConfig`

Reply:
I understand the maxim of composition over inheritance, but personally I value codebase consistency higher.
Do you have a more specific justification in this case that would motivate us to migrate all other AbstractConfig subclasses to composition?

After all, the AbstractConfig documentation promotes this style: https://github.com/apache/kafka/blob/f2aafcc66faca4d07a27e4ef90158136cb317b44/clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java#L41

Reply:
> I value codebase consistency more.
> After all, the AbstractConfig documentation promotes this style: 

ya, that is a good reason. However, the rule is written in 2014 and I don't think "to extend" means "can't compose"

> Do you have a more specific justification in this case that would motivate us to migrate all other AbstractConfig subclasses to composition?

that is not what I try to say. We don't need to change whole code base when we are aware of good pattern, but it would be nice to apply on the new code.

the design pattern is a long story. I love to have more discussion, but it is not worth obstructing the progress from this PR. 

Please ignore my comments and let's move on this PR :)

Reply:
> However, the rule is written in 2014 and I don't think "to extend" means "can't compose"

I completely agree. There are lots of decisions from 2014/2015 that I disagree with and actively cause me problems,  this one fortunately hasn't been one of them.

> We don't need to change whole code base when we are aware of good pattern, but it would be nice to apply on the new code.

I agree that the two styles can coexist at the same time, and implementing a pattern on "new" code is a good rollout plan. I do think that one should eventually "win", and I was curious if you had a good reason that might make composition win :)

Reply:
> I was curious if you had a good reason that might make composition win :)

There are two main reasons.

1. Extending the `AbstractConfig` will have a bunch of "public" APIs. Those unused public APIs  will make subclass verbose in using
2. the exposed APIs may mislead callers to take "unrelated" configs from  `RemoteLogManagerConfig`. And that may cause more consistency in the future